src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/MvcDetector.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure;
+
+internal static class MvcDetector
+{
+    private const string ControllerTypeNameSuffix = "Controller";
+
+    // Replicates logic from ControllerFeatureProvider.IsController.
+    // https://github.com/dotnet/aspnetcore/blob/785cf9bd845a8d28dce3a079c4fedf4a4c2afe57/src/Mvc/Mvc.Core/src/Controllers/ControllerFeatureProvider.cs#L39
+    public static bool IsController(ITypeSymbol typeSymbol, SemanticModel semanticModel)
```

Comment:
either take in an INamedTypeSymbol, or immediately bail if you get a non INamedTypeSymbol.  (unless you genuinely do want this working for things like type parameters).